lib/parse_identifier.rb

```diff
@@ -0,0 +1,121 @@
+require 'identifiers'
+
+# https://github.com/altmetric/identifiers
+#
+# Identifiers::DOI.extract('example: 10.123/abcd.efghi')
+# # => ["10.123/abcd.efghi"]
+#
+# It can support many identifiers:
+# Identifiers::AdsBibcode.extract('')
+# Identifiers::ArxivId.extract('')
+# Identifiers::DOI.extract('')
+# Identifiers::Handle.extract('')
+# Identifiers::ISBN.extract('')
+# Identifiers::NationalClinicalTrialId.extract('')
+# Identifiers::ORCID.extract('')
+# Identifiers::PubmedId.extract('')
+# Identifiers::RepecId.extract('')
+# Identifiers::URN.extract('')
+
+# Exception for incorrect instantiation of a parser (not necessarily bad data)
+class ParseIdentifierTypeError < StandardError
+
+end
+
+# Exception for bad data
+class ParseIdentifierInvalidError < StandardError
+
+end
+
+# Parse the identifiers in a PublicationIdentifier
+#
+# This is a base class for ParseIdentifier* subclasses.
+#
+# This class should try to work with the input parameter as a read-only object.
+# It should only modify it when explicitly asked to `update` it.  If it's
+# an active record object, assign new values to it, but don't save it, leave
+# that persistence action to the consumer; see the runner script in
+# @see script/publication_identifier_normalization.rb
+class ParseIdentifier
+
+  URI_PREFIX = nil # subclasses override this constant
```

Comment:
That is the wrong pattern for a `constant`.  Logic in this class should be able to rely on the `constant` being the value that is set.  Logic in subclasses should not "override" the constant.  That OO pattern is for *methods*.

Reply:
What is wrong with overriding a constant?  It also works.  I've not seen any dictum that overriding a constant is a bad thing.  If so, point me to the dictum(s)?

Reply:
Removed.